File: [127]单词接龙.cpp
#include <string>
#include <vector>
#include <map>
#include <set>
#include <queue>
using namespace std;

class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        map<string, set<string> > comboDict;
        map<string, bool> visited;
        queue<string> q;
        q.push(beginWord);
        int word_length = beginWord.length();
        int ans = 1;

        for (string word : wordList) {
            visited[word] = false;
            for (int i = 0 ; i < word_length; ++i) {
                string newWord = word.substr(0, i) + '*' + word.substr(i + 1, word_length);
                comboDict[newWord].insert(word);
            }
        }
        visited[beginWord] = true;
        
        while (!q.empty()) {
            int q_size = q.size();
            ++ans;
            for (int i = 0; i < q_size; ++i) {
                string word = q.front(); q.pop();
                for (int j = 0; j < word_length; ++j) {
                    string newWord = word.substr(0, j) + '*' + word.substr(j + 1, word_length);
                    for (string str : comboDict[newWord]) {
                        if (str == endWord){
                            return ans;
                        }

                        if (!visited[str]) {
                            visited[str] = true;
                            q.push(str);
                        }
                    }
                }
            }
        }
        return 0;
    }
};

File: [127]单词接龙.py
from itertools import product


class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        is_used = {s:False for s in wordList}

        def has_edge(pre_word, pro_word):
            count = 0
            for i in range(len(pre_word)):
                count += ( 1 if pre_word[i]!=pro_word[i] else 0)
                if count > 1:
                    return False
            # print(pre_word, pro_word, (not is_used[pro_word] and count == 1))
            return count == 1

        queue = [beginWord]
        count = 0
        has_found = False
        while not has_found and queue:
            length = len(queue)
            for _ in range(length):
                word = queue.pop(0)
                if (word == endWord):
                    has_found = True
                    break

                for pro_word in wordList:
                    if not is_used[pro_word] and has_edge(word, pro_word):
                        queue.append(pro_word)
                        is_used[pro_word] = True
                
            count += 1
        
        return count if has_found else 0
